controllers/secret_controller.go

```diff
@@ -98,53 +112,112 @@ var defaultNamespaces = []string{
+	if err != nil {
+		reqLogger.Error(err, "Error determining cluster readiness.")
+		return r.Readiness.Result(), err
+	}
+
+	// Get a list of all relevant objects in the `openshift-monitoring` namespace.
+	// This is used for determining which secrets and configMaps are present so that the necessary
+	// Alertmanager config changes can happen later.
+	opts := []client.ListOption{
+		client.InNamespace(request.Namespace),
+	}
+	secretList := &corev1.SecretList{}
+	err = r.Client.List(context.TODO(), secretList, opts...)
+	if err != nil {
+		reqLogger.Error(err, "Unable to list secrets")
 	}
-}
 
-// add adds a new Controller to mgr with r as the reconcile.Reconciler
-func add(mgr manager.Manager, r reconcile.Reconciler) error {
-	// Create a new controller
-	c, err := controller.New("secret-controller", mgr, controller.Options{Reconciler: r})
+	cmList := &corev1.ConfigMapList{}
+	err = r.Client.List(context.TODO(), cmList, opts...)
 	if err != nil {
-		return err
+		reqLogger.Error(err, "Unable to list configMaps")
 	}
 
-	// Watch for changes to primary resource (type "Secret").
-	// For each Add/Update/Delete event, the reconcile loop will be sent a reconcile Request.
-	err = c.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForObject{})
+	pagerdutyRoutingKey, watchdogURL := r.parseSecrets(reqLogger, secretList, request.Namespace, clusterReady)
+	osdNamespaces := r.parseConfigMaps(reqLogger, cmList, request.Namespace)
+	reqLogger.Info("DEBUG: Adding PagerDuty routes for the following namespaces", "Namespaces", osdNamespaces)
+
+	ocmAgentURL := r.readOCMAgentServiceURLFromConfig(reqLogger, cmList, request.Namespace)
+
+	clusterProxy, err := r.getClusterProxy()
 	if err != nil {
-		return err
+		reqLogger.Error(err, "Unable to get cluster proxy")
 	}
 
-	err = c.Watch(&source.Kind{Type: &corev1.ConfigMap{}}, &handler.EnqueueRequestForObject{})
+	// create the desired alertmanager Config
+	clusterID, err := r.getClusterID()
 	if err != nil {
-		return err
+		reqLogger.Error(err, "Error reading cluster id.")
 	}
-	return nil
+	alertmanagerconfig := createAlertManagerConfig(pagerdutyRoutingKey, watchdogURL, ocmAgentURL, clusterID, clusterProxy, osdNamespaces)
+
+	// write the alertmanager Config
+	writeAlertManagerConfig(r, reqLogger, alertmanagerconfig)
+
+	// Update metrics after all reconcile operations are complete.
+	metrics.UpdateSecretsMetrics(secretList, alertmanagerconfig)
+	metrics.UpdateConfigMapMetrics(cmList)
+	reqLogger.Info("Finished reconcile for secret.")
+
+	// The readiness Result decides whether we should requeue, effectively "polling" the readiness logic.
+	return r.Readiness.Result(), nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	client := mgr.GetClient()
+
+	r.Readiness = &readiness.Impl{Client: client}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.ConfigMap{}).
```

Comment:
With this being a _secret_ controller it may be more correct to have this as a controller `For` Secrets (its primary resource) which also `Watches` ConfigMaps, rather than the other way around.

Reply:
Thanks Matt. That does make more sense.
I will get that updated.

Reply:
Thanks! The fix looks good.